include/boost/math/statistics/univariate_statistics.hpp

```diff
@@ -512,6 +512,74 @@ inline auto interquartile_range(RandomAccessContainer & v)
     return interquartile_range(v.begin(), v.end());
 }
 
+template<class ForwardIterator, class OutputIterator>
+auto sorted_mode(ForwardIterator first, ForwardIterator last, OutputIterator output) -> decltype(output)
+{
+    using Z = typename std::iterator_traits<ForwardIterator>::value_type;
+    static_assert(std::is_integral<Z>::value, "Floating point values have not yet been implemented.");
+    using Size = typename std::iterator_traits<ForwardIterator>::difference_type;
+
+    std::vector<Z> modes {};
+    modes.reserve(16);
+    Size max_counter {0};
+    auto it {first};
```

Comment:
It's perfectly OK to use `first` as the variable that gets modified, you don't have to make a copy of it. That's generally how STL algorithms are implemented. `while (first != last)` was Stepanov's mantra.